extensions/core/src/vtabs.rs

```diff
@@ -103,3 +135,172 @@ pub trait VTabCursor: Sized {
+            idx_num: self.idx_num,
+            estimated_cost: self.estimated_cost,
+            estimated_rows: self.estimated_rows,
+            order_by_consumed: self.order_by_consumed,
+            constraint_usages_ptr: ptr,
+            constraint_usage_len: len,
+            idx_str: c_idx_str as *mut _,
+            idx_str_len,
+        }
+    }
+
+    /// Reclaims ownership of `constraint_usages` and `idx_str` from an FFI-safe `ExtIndexInfo`.
+    /// # Safety
+    /// This method is unsafe because it can cause memory leaks if not used correctly.
+    /// to_ffi and from_ffi are meant to send index info across ffi bounds then immediately reclaim it.
+    pub unsafe fn from_ffi(ffi: ExtIndexInfo) -> Self {
+        let constraint_usages = unsafe {
+            Box::from_raw(std::slice::from_raw_parts_mut(
+                ffi.constraint_usages_ptr,
+                ffi.constraint_usage_len,
+            ))
+            .to_vec()
+        };
+        let idx_str = if ffi.idx_str.is_null() {
+            None
+        } else {
+            Some(unsafe {
+                std::ffi::CString::from_raw(ffi.idx_str as *mut _)
+                    .to_string_lossy()
+                    .into_owned()
+            })
+        };
+        Self {
+            idx_num: ffi.idx_num,
+            idx_str,
+            order_by_consumed: ffi.order_by_consumed,
+            estimated_cost: ffi.estimated_cost,
+            estimated_rows: ffi.estimated_rows,
+            constraint_usages,
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, Debug)]
+/// FFI representation of IndexInfo.
+pub struct ExtIndexInfo {
+    pub idx_num: i32,
+    pub idx_str: *const u8,
+    pub idx_str_len: usize,
+    pub order_by_consumed: bool,
+    pub estimated_cost: f64,
+    pub estimated_rows: u32,
+    pub constraint_usages_ptr: *mut ConstraintUsage,
+    pub constraint_usage_len: usize,
+}
+
+/// Returned from xBestIndex to describe how the virtual table
+/// can use the constraints in the WHERE clause of a query.
+#[derive(Debug, Clone, Copy)]
+pub struct ConstraintUsage {
+    /// 1 based index of the argument passed
+    pub argv_index: Option<u32>,
+    /// If true, core can omit this constraint in the vdbe layer.
+    pub omit: bool,
+}
+
+#[derive(Clone, Copy, Debug)]
+#[repr(C)]
+/// The primary argument to xBestIndex, which describes a constraint
+/// in a query involving a virtual table.
+pub struct ConstraintInfo {
+    /// The index of the column referenced in the WHERE clause.
+    pub column_index: u32,
+    /// The operator used in the clause.
+    pub op: ConstraintOp,
+    /// Whether or not constraint is garaunteed to be enforced.
+    pub usable: bool,
+    /// packed integer with the index of the constraint in the planner,
+    /// and the side of the binary expr that the relevant column is on.
+    pub plan_info: u32,
+}
+
+impl ConstraintInfo {
+    #[inline(always)]
+    pub fn pack_plan_info(pred_idx: u32, is_right_side: bool) -> u32 {
```

Comment:
heh, cool

Reply:
very similar `struct IndexConstraint` elsewhere in the planner, without all the packing